### PR TITLE
Support for custom distance functions

### PR DESCRIPTION
**Pull request by MasterEric**
As mentioned in an aside in issue #1. This pull request provides the following improvements:

* Drastically expands documentation, for ease of maintenance. and code readability.
* Removed the unused internal function `size()`.
* Adds performance optimizations.
  * Identity distances are no longer calculated (distance from a point to itself).
  * Halves the number of distance calculations performed by eliminating redundant calculations.
* Adds support for a custom distance function.
  * Defaults to the current distance function, which has been renamed to `euclidean()`.
  * Accepts two Points and returns a Number distance.
  * Allows for customization of how distances are handled, useful in cases such as lat/lng values.
  * Interesting test case: set the distance function to `(x,y) => 1/euclidean(x,y)` to see the function optimize for the longest path.

Pinging the maintainer @lovasoa, as well as interested users @spridev and @mariusa.

### CODE REVIEW

File: salesman.js [hunk 1]
```diff
@@ -10,35 +11,80 @@
...
+ * Perform one iteration of the simulated annealing.
+ * 
+ * Choose two random points in the path, and calculate how much the path distance would change
+ * if you swapped the two points. If it would make the path shorter, swap them.
+ * 
+ * If not, have a random chance to swap them anyway.
+ * This random chance is based on how bad the move is,
+ * as well as how early in the annealing process we are (the "temperature").
+ * 
+ * @param {*} temp The current temperature of the algorithm.
```
Comment (lovasoa) [inline salesman.js:58]:
```suggestion
 * @param {number} temp The current temperature of the algorithm.
```

File: salesman.js [hunk 2]
```diff
@@ -10,35 +11,80 @@
...
-  var s = 0;
-  for (var i=0; i<this.points.length; i++) {
-    s += this.distance(i, ((i+1)%this.points.length));
-  }
-  return s;
-};
+/**
+ * Swap two points in the path order by their indices.
+ * @param {*} i The first index to swap.
+ * @param {*} j The second index to swap.
```
Comment (lovasoa) [inline salesman.js:70]:
```suggestion
 * @param {number} i The first index to swap.
 * @param {number} j The second index to swap.
```

File: salesman.js [hunk 3]
```diff
@@ -10,35 +11,80 @@
...
+/**
+ * Calculate the change in path distance if i and j were swapped.
+ * 
+ * Calculate the distance between i and j's neighbors,
+ * plus the distance between j and i's neighbors, minus the current distances.
+ * 
+ * If the value is negative, it would make the path shorter to swap the values.
+ * @param {*} i The first index to compare.
+ * @param {*} j The second index to compare.
+ * @returns The change in path distance if i and j were swapped.
```
Comment (lovasoa) [inline salesman.js:86]:
```suggestion
 * @param {number} i The first index to compare.
 * @param {number} j The second index to compare.
 * @returns {number} The change in path distance if i and j were swapped.
```

File: salesman.js [hunk 4]
```diff
@@ -57,29 +103,65 @@ Path.prototype.delta_distance = function(i, j) {
...
+ */
 Path.prototype.distance = function(i, j) {
-  return this.distances[this.order[i] * this.points.length + this.order[j]];
+  if (i === j) return 0; // Identity.
+
+  // Ensure low is actually lower.
+  var low = this.order[i], high = this.order[j];
+  if (low > high) { low = this.order[j]; high = this.order[i]; }
+
+  return this.distances[low * this.points.length + high] || 0;
```
Comment (lovasoa) [inline salesman.js:134]: Is this really faster then what we had ? This looks much more branchy.

File: .gitignore [hunk 5]
```diff
@@ -35,3 +35,5 @@ jspm_packages
 
 # Optional REPL history
 .node_repl_history
+
+package-lock.json
```
Comment (lovasoa) [inline .gitignore:39]:
```suggestion
```

https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json

**Review by lovasoa — CHANGES_REQUESTED** [review]
Thank you very much for your contribution ! This is very welcome, especially the added documentation.

A few remarks:
 - Some type annotations are missing or specify the `Number` class instead of the `number` base type
 - The demo UI changes should be in a separate pull request, against the gh-pages branch
 - Did you measure the performance impact of your changes ? Are you sure the added branches on the hot path don't make the code slower instead of making it faster ?

**Comment by MasterEric** [comment]
@lovasoa I've made the suggested improvements.

Additionally, I've created a short script for testing purposes. This script creates 5,000 points in a 100x100 grid, and creates a path, then repeats this process 500 times, then outputs the average, maximum, and minimum duration taken per iteration.

Here were the results I got. If you want more concrete results, a lengthier test might be desirable (simply change the variables at the top).

```
BEFORE CHANGES:
* Average Time: 0.23145491981220262
* Max Time: 0.6108302001953125
* Min Time: 0.197708899974823
```

```
AFTER CHANGES:
* Average Time: 0.24901885180091873
* Max Time: 0.331204400062561
* Min Time: 0.21017019987106322
```

It appears that the times have become more consistent after the changes. The average time and minimum times have increased slightly but the maximum time has reduced by almost half.

Or maybe there's some flaw in the test?

File: .gitignore [hunk 6]
```diff
@@ -1,37 +1,37 @@
...
+*.log
+npm-debug.log*
+
+# Runtime data
+pids
+*.pid
+*.seed
+
+# Directory for instrumented libs generated by jscoverage/JSCover
+lib-cov
```
Comment (lovasoa) [inline .gitignore:12]: What did you change here ?

File: README.md [hunk 7]
```diff
@@ -3,14 +3,15 @@
...
+
 salesman npm module
 
 Good heuristic for the traveling salesman problem using simulated annealing.  
 
 * [salesman](#module_salesman)
     * [~Point](#module_salesman..Point)
         * [new Point(x, y)](#new_module_salesman..Point_new)
-    * [~solve(points, [temp_coeff], [callback=])](#module_salesman..solve) ⇒ <code>Array.&lt;Number&gt;</code>
+    * [~solve(points, [temp_coeff], [callback], [callback])](#module_salesman..solve) ⇒ <code>[ &#x27;Array&#x27; ].&lt;Number&gt;</code>
```
Comment (lovasoa) [inline README.md:14]: Why `['Array']` ?

**Comment by lovasoa** [comment]
I plotted the test results. It looks like the performance indeed improved 🎉 

image

[perf.zip](https://github.com/lovasoa/salesman.js/files/5671672/perf.zip)

File: salesman.js (hunk 1 above)
Comment (lovasoa) [reply salesman.js:58]: So, you didn't apply the suggestion ?

File: README.md (hunk 7 above)
Comment (MasterEric) [reply README.md]: I didn't do that, jsdoc-to-markdown created that file.

**Comment by Deses** [comment]
This performance improvement was left in limbo? is there any chance you can merge it?

**Comment by lovasoa** [comment]
Better late than never 😑
Sorry for the 2 year Friday

**Pull request merged**